src/Thorn_Calculus-Core_Bisimilarities.thy

```diff
@@ -1350,9 +1350,148 @@ text \<open>
+    have "A \<triangleright> \<star>\<^bsup>n\<^esup> X = IO \<eta> A' n X" and "Q' = post_receive n X (\<lambda>v. \<P> v \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x)"
+      by (fastforce elim: repeated_receive_evolution)+
+    with \<open>A \<triangleright>\<^sup>\<infinity> x. \<P> x \<rightarrow>\<^sub>s\<lparr>IO \<eta> A' n X\<rparr> Q'\<close> and \<open>\<alpha> = IO \<eta> A' n X\<close>
+    have "A \<triangleright>\<^sup>\<infinity> x. \<P> x \<rightarrow>\<^sub>s\<lparr>A \<triangleright> \<star>\<^bsup>n\<^esup> X\<rparr> post_receive n X (\<lambda>v. \<P> v \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x)" and \<open>\<alpha> = A \<triangleright> \<star>\<^bsup>n\<^esup> X\<close>
+      by (simp_all only:)
+    moreover
+    have "
+      (A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n \<parallel> post_receive n X (\<lambda>v. \<P> v \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x)
+      \<sim>\<^sub>s
+      post_receive n X \<P> \<parallel> (A \<triangleright>\<^sup>\<infinity> x. \<P> x \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n"
+    proof -
+      have "
+        post_receive n X (\<lambda>v. \<P> v \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x)
+        =
+        post_receive n X \<P> \<parallel> (A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n"
+        using post_receive_after_parallel
+        by simp
+      then have "
+        (A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n \<parallel> post_receive n X (\<lambda>v. \<P> v \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x)
+        \<sim>\<^sub>s
+        post_receive n X \<P> \<parallel> ((A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n \<parallel> (A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n)"
+        by (simp only: parallel_left_commutativity)
+      then show ?thesis
+        by (simp only: adapted_after_parallel)
+    qed
+    ultimately show ?thesis
+      unfolding
+        \<open>\<alpha> = A \<triangleright> \<star>\<^bsup>n\<^esup> X\<close> and \<open>Q' = post_receive n X (\<lambda>v. \<P> v \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x)\<close> and
+        \<open>S = (A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n \<parallel> Q'\<close>
+      using
+        post_receive_after_parallel and
+        composition_in_universe [
+          OF
+            suffix_adapted_mutation_in_universe
+            parallel_mutation_in_universe
+        ]
+      by (intro exI conjI, use in assumption) (fastforce intro: rev_bexI)
+  qed (blast elim: repeated_receive_evolution(1))+
+next
+  case (backward_simulation \<alpha> S)
+  then show ?case
+  proof (cases rule: repeated_receive_evolution [case_names receiving])
```

Comment:
I think it would be better to use `repeated_receive_evolution` with a proof method to show what holds in the one case possible (like you did under `forward_simulation`) instead of using `cases`. That would make the code more straightforward (as it would avoid the extra nesting with just a single case) and would make it consistent with the code for `forward_simulation`.

Reply:
Do you mean something like the following?
```isabelle
  case (backward_simulation α S)
  then show ?case
  proof cases
      case (receiving n X)
      ... (* proof for `receiving`, no other case in explicitly mentioned *)
   qed (blast elim: transition_from_repeated_receive(1))+
```

Reply:
Please ignore my previous comment, I hadn't fully understood what your suggestion was. Fixed in 4c711e10243ea6a6b7394c40245a635f3efe9a85, please verify.

Reply:
I fixed a minor style issue, introduced by 4c711e10243ea6a6b7394c40245a635f3efe9a85, in efb2d9722930d497e4c555c143792c6e4da4a785.

Reply:
Hmm, I can’t quickly assess these changes, given that they are shown in GitHub using this LaTeX-like encoding, but since we went together through the whole proof and came up with 04d4c63b3adfd44788c698dcdb9b05b694c12814, I think everything should be fine.